fcd/pass_argrec_remill.cpp

```diff
@@ -0,0 +1,369 @@
+  std::vector<llvm::Type *> params;
+  for (auto reg : used_regs) {
+    auto var = remill::FindVarInFunction(func, reg);
+    auto inst = llvm::dyn_cast<llvm::AllocaInst>(var);
+    CHECK(inst != nullptr);
+    auto type = llvm::dyn_cast<llvm::PointerType>(inst->getAllocatedType());
+    CHECK(type != nullptr);
+    params.push_back(type->getElementType());
+  }
+
+  auto ret = RecoverRetType(func, cc);
+
+  std::stringstream cc_func_name;
+  cc_func_name << sPrefix << func->getName().str();
+
+  auto cc_type = llvm::FunctionType::get(ret, params, false);
+  auto cc_func = llvm::dyn_cast<llvm::Function>(
+      func->getParent()->getOrInsertFunction(cc_func_name.str(), cc_type));
+
+  CHECK(cc_func != nullptr);
+
+  for (auto &arg : cc_func->args()) {
+    std::stringstream cc_arg_name;
+    cc_arg_name << sPrefix << used_regs[arg.getArgNo()];
+    arg.setName(cc_arg_name.str());
+  }
+
+  return cc_func;
+}
+
+static void StoreRegArgsToLocals(llvm::Function *func) {
+  llvm::IRBuilder<> ir(func->getEntryBlock().getTerminator());
+  for (auto &arg : func->args()) {
+    auto name = TrimPrefix(arg.getName().str());
+    auto var = remill::FindVarInFunction(func, name);
+    auto ptr = ir.CreateLoad(var);
+    ir.CreateStore(&arg, ptr);
+  }
+}
+
+static void ConvertRemillArgsToLocals(llvm::Function *func) {
+  auto module = func->getParent();
+  auto &entry = func->getEntryBlock();
+
+  llvm::IRBuilder<> ir(&entry, entry.begin());
+
+  auto mem_type = remill::MemoryPointerType(module)->getElementType();
+  auto arg_mem = remill::NthArgument(func, remill::kMemoryPointerArgNum);
+  auto loc_mem = ir.CreateAlloca(mem_type, nullptr, "loc_memory");
+  arg_mem->replaceAllUsesWith(loc_mem);
+
+  ir.SetInsertPoint(loc_mem);
+
+  auto pc_type = remill::AddressType(module);
+  auto arg_pc = remill::NthArgument(func, remill::kPCArgNum);
+  auto loc_pc = ir.CreateAlloca(pc_type, nullptr, "loc_pc");
+  arg_pc->replaceAllUsesWith(loc_pc);
+
+  ir.SetInsertPoint(loc_pc);
+
+  auto state_type = remill::StatePointerType(module)->getElementType();
+  auto arg_state = remill::NthArgument(func, remill::kStatePointerArgNum);
+  auto loc_state = ir.CreateAlloca(state_type, nullptr, "loc_state");
+  arg_state->replaceAllUsesWith(loc_state);
+}
+
+static bool IsLiftedFunction(llvm::Function *func) {
+  auto module = func->getParent();
+  if (func->getFunctionType() == remill::LiftedFunctionType(module)) {
+    if (!func->getName().startswith("__remill")) {
+      return !func->empty();
+    }
+  }
+
+  return false;
+}
+
+}  // namespace
+
+char RemillArgumentRecovery::ID = 0;
+
+RemillArgumentRecovery::RemillArgumentRecovery()
+    : ModulePass(RemillArgumentRecovery::ID),
+      cc(CallingConvention(remill::GetTargetArch()->DefaultCallingConv())) {
+  sPrefix = cPrefix;
+}
+
+void RemillArgumentRecovery::getAnalysisUsage(
+    llvm::AnalysisUsage &usage) const {}
+
+bool RemillArgumentRecovery::runOnModule(llvm::Module &module) {
```

Comment:
Can you add a bit more commenting in this file to describe the gist of what's going on?

Reply:
Addressed in 01a78f6. Also added comments to `RemillStackRecovery` to clarify the steps of the algorithm.